keps/sig-node/2570-memory-qos/README.md

```diff
@@ -0,0 +1,171 @@
+# KEP-2570: Support Memory QoS using cgroups v2
+## Summary
+Support memory qos in pod qos class using cgroups v2.
+
+## Motivation
+In traditional cgroups v1 implement in Kubernetes, pod qos class and qos manager can only limit cpu resources, such as cpu_shares, memory qos is not yet implemented. cgroups v2 brings new capabilities for memory controller and it would help Kubernetes enhance memory isolation quality.
+
+### Goals
+- Support memory qos for pod qos classes
+
+### Non-Goals
+- additional qos design
+- support other resources qos
+
+## Proposal
+Use memory controller interfaces of cgroups v2 to support memory qos of pod qos class.
+
+Kubernetes cgroup manager uses `memory.limit_in_bytes` in cgroups v1 and `memory.max` in cgroups v2 to limit memory capacity and uses `oom_scores` to determine order of killing container process when OOM occurs.
+
+Above implement has following disadvantages:
+- `Guaranteed` pod memory can not be fully reserved, page cache is at risk of being recycled
+- `Burstable` pod memory allocation is not well protected, there may occur allocation latency frequently when node memory nearly runs out
+
+Currently we only use `memory.limit_in_bytes = sum(pod.spec.containers.resources.limits[memory])` to limit memory hardly in cgroups v1 and `memory.max = sum(pod.spec.containers.resources.limits[memory])` in cgroups v2. And `resources.requests[memory]` is not yet used neither cgroups v1 nor cgroups v2.
+
+cgroups v2 introduces a better way to limit and guarantee memory quality.
+
+| File | Description |
+| -------- | -------- |
+| memory.high | memory.high is the memory usage throttle limit. This is the main mechanism to control a cgroup’s memory use. If a cgroup's memory use goes over the high boundary specified here, the cgroup’s processes are throttled and put under heavy reclaim pressure. The default is max, meaning there is no limit. |
+| memory.max | memory.max is the memory usage hard limit, acting as the final protection mechanism: If a cgroup's memory usage reaches this limit and can't be reduced, the system OOM killer is invoked on the cgroup. Under certain circumstances, usage may go over the memory.high limit temporarily. When the high limit is used and monitored properly, memory.max serves mainly to provide the final safety net. The default is max. |
+| memory.low | memory.low is the best-effort memory protection, a “soft guarantee” that if the cgroup and all its descendants are below this threshold, the cgroup's memory won't be reclaimed unless memory can’t be reclaimed from any unprotected cgroups. |
```

Comment:
Can we capture that the memory reclaim pressure is proportional to the overage for memory.low and memory.min?

Reply:
I think we can calculate the relative reclaim pressure times. e.g. if two peer cgroups A and B are currently at 3G, but A has 2G of memory.low protection while B doesn't, B will experience three times higher reclaim pressure.

Reply:
Yes, that's what I want to make sure we capture here for reviewers :)